rpc-server/src/main.rs

```diff
@@ -19,52 +31,250 @@ mod utils;
+/// be parsed (using [`RpcRequest::parse`]) from the `request.params`.  Ok
+/// results of the `callback` will be converted into a [`Value`] via serde
+/// serialisation.
+async fn process_method_call<R, V, E, F>(
+    request: Request,
+    callback: impl FnOnce(R) -> F,
+) -> Result<Value, RPCError>
 where
-    R: Router,
+    R: RpcRequest,
+    V: serde::ser::Serialize,
+    RPCError: std::convert::From<E>,
+    F: std::future::Future<Output = Result<V, E>>,
 {
-    fn with_method_and_metrics<F, T, P>(self, name: &'static str, handler: &'static F) -> Self
-    where
-        F: jsonrpc_v2::Factory<T, errors::RPCError, (Data<config::ServerContext>, Params<P>)>
-            + Send
-            + Sync
-            + 'static,
-        T: serde::Serialize + Send + 'static,
-        P: serde::de::DeserializeOwned + Send + 'static,
-    {
-        self.with_method(name, |data, params| async {
-            let result = handler.call((data, params)).await;
+    serialize_response(callback(R::parse(request.params)?).await?)
+}
 
-            if let Err(err) = &result {
-                match &err.error_struct {
-                    Some(near_jsonrpc::primitives::errors::RpcErrorKind::RequestValidationError(
-                        near_jsonrpc::primitives::errors::RpcRequestValidationErrorKind::ParseError { .. },
-                    )) => {
-                        metrics::METHOD_ERRORS_TOTAL
-                            .with_label_values(&[name, "PARSE_ERROR"])
-                            .inc();
-                    }
-                    Some(near_jsonrpc::primitives::errors::RpcErrorKind::HandlerError(error_struct)) => {
-                        if let Some(stringified_error_name) = error_struct.get("name").and_then(serde_json::Value::as_str) {
-                            metrics::METHOD_ERRORS_TOTAL
-                                .with_label_values(&[name, stringified_error_name])
-                                .inc();
-                        }
-                    }
-                    Some(near_jsonrpc::primitives::errors::RpcErrorKind::InternalError(_)) => {
-                        metrics::METHOD_ERRORS_TOTAL
-                            .with_label_values(&[name, "INTERNAL_ERROR"])
-                            .inc();
+async fn rpc_handler(
+    data: web::Data<config::ServerContext>,
+    payload: web::Json<Message>,
+) -> HttpResponse {
+    let Message::Request(request) = payload.0 else {
+        return HttpResponse::BadRequest().finish();
+    };
+
+    let id = request.id.clone();
+
+    let method_name = request.method.clone();
+    let result = match method_name.as_ref() {
+        // custom request methods
+        "view_state_paginated" => {
+            if let Ok(request_data) = serde_json::from_value(request.params) {
+                serialize_response(
+                    modules::state::methods::view_state_paginated(data, request_data).await,
```

Comment:
`RpcViewStatePaginatedRequest` doesn't support `RpcRequest` trait and it cannot be parsed. Also we can't simply implement this trait now, because it requires `Params` structure to be accessible from read-rpc. Unfortunately, for now this is not the case as `Params` structure is declared as `pub(crate)`, so we can't use it for now. So we either need to wait until my PR related to splitting jsonrpc into public trait will be merged (because I've moved `Params` structure there and it will be public to everybody) or think about another solution